File: code/comfyui/worker/java/src/main/java/dev/learn/comfy/worker/ResultStore.java
package dev.learn.comfy.worker;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.time.Instant;
import java.util.HexFormat;
import java.util.List;

import dev.learn.comfy.worker.Jobs.StoredFile;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.ObjectNode;

/**
 * Results on a shared directory, one folder per job id. done.json says a job is finished; a claim file, created
 * with CREATE_NEW, lets only one worker run a job at a time. In production: object storage and a unique database key.
 */
public final class ResultStore {
    private final Path root;

    public ResultStore(Path root) {
        this.root = root;
    }

    public Path root() {
        return root;
    }

    public boolean isDone(String jobId) {
        return Files.exists(root.resolve(jobId).resolve("done.json"));
    }

    /** True if this worker now owns the job. A claim older than its lease is taken over: its worker died. */
    public boolean tryClaim(String jobId, String owner, Duration lease) throws IOException {
        Files.createDirectories(root.resolve(jobId));
        Path path = root.resolve(jobId).resolve("claim");
        for (int tries = 0; tries < 2; tries++) {
            try {
                ObjectNode claim = ComfyInstance.JSON.createObjectNode().put("owner", owner).put("expires", Instant.now().plus(lease).toString());
                // CREATE_NEW fails if the file exists, atomically.
                Files.writeString(path, claim.toString(), StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
                return true;
            } catch (FileAlreadyExistsException e) {
                JsonNode claim = readClaim(path);
                if (claim == null || Instant.parse(claim.path("expires").asString()).isAfter(Instant.now())) {
                    return false;
                }
                Files.deleteIfExists(path); // expired: try once more
            }
        }
        return false;
    }

    private static JsonNode readClaim(Path path) {
        try {
            JsonNode claim = ComfyInstance.JSON.readTree(Files.readString(path));
            return claim.has("expires") ? claim : null;
        } catch (IOException | JacksonException e) {
            return null; // being written by its owner, or already gone
        }
    }

    public void releaseClaim(String jobId, String owner) throws IOException {
        Path path = root.resolve(jobId).resolve("claim");
        JsonNode claim = readClaim(path);
        if (claim != null && claim.path("owner").asString().equals(owner)) {
            try {
                Files.delete(path);
            } catch (NoSuchFileException e) {
                // already gone
            }
        }
    }

    /** Writes an output file under a temporary name first: a crash never leaves a half file that looks finished. */
    public StoredFile saveFile(String jobId, String node, String name, byte[] bytes) throws IOException {
        Path dir = Files.createDirectories(root.resolve(jobId).resolve("outputs"));
        Path fileName = Path.of(name).getFileName();
        Path target = dir.resolve(node + "-" + fileName);
        Path temp = dir.resolve(target.getFileName() + ".partial");
        Files.write(temp, bytes);
        Files.move(temp, target, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
        return new StoredFile(node, target.getFileName().toString(), bytes.length, sha256(bytes).substring(0, 16));
    }

    public void complete(String jobId, String owner, String instance, int attempt, List<StoredFile> files) throws IOException {
        ObjectNode done = ComfyInstance.JSON.createObjectNode().put("job", jobId).put("worker", owner).put("gpu", instance).put("attempts", attempt);
        var array = done.putArray("files");
        for (StoredFile file : files) {
            array.addObject().put("node", file.node()).put("name", file.name()).put("bytes", file.bytes()).put("sha256", file.sha256());
        }
        Path path = root.resolve(jobId).resolve("done.json");
        Path temp = root.resolve(jobId).resolve("done.json.partial");
        Files.writeString(temp, ComfyInstance.JSON.writerWithDefaultPrettyPrinter().writeValueAsString(done), StandardCharsets.UTF_8);
        Files.move(temp, path, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
        releaseClaim(jobId, owner);
    }

    public void fail(String jobId, String owner, String reason) throws IOException {
        Files.writeString(root.resolve(jobId).resolve("failed.txt"), reason + System.lineSeparator(), StandardCharsets.UTF_8);
        releaseClaim(jobId, owner);
    }

    static String sha256(byte[] bytes) {
        try {
            return HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(bytes));
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: code/comfyui/worker/csharp/ComfyWorker/ResultStore.cs
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Learn.Comfy.Worker;

/// <summary>
/// Results on a shared directory, one folder per job id. It is also what makes the worker idempotent:
/// done.json says a job is finished, and a claim file, created with FileMode.CreateNew, lets only one
/// worker run a job at a time. In production this would be object storage and a database row with a unique key.
/// </summary>
public sealed class ResultStore(string root)
{
    public string Root => root;

    string JobDir(string jobId) => Path.Combine(root, jobId);

    public bool IsDone(string jobId) => File.Exists(Path.Combine(JobDir(jobId), "done.json"));

    /// <summary>True if this worker now owns the job. A claim older than its lease is taken over: its worker died.</summary>
    public bool TryClaim(string jobId, string owner, TimeSpan lease)
    {
        Directory.CreateDirectory(JobDir(jobId));
        string path = Path.Combine(JobDir(jobId), "claim");
        for (int tries = 0; tries < 2; tries++)
        {
            try
            {
                // CreateNew fails if the file exists, atomically, on every OS and on most network file systems.
                using var file = new FileStream(path, FileMode.CreateNew, FileAccess.Write);
                JsonSerializer.Serialize(file, new JsonObject { ["owner"] = owner, ["expires"] = DateTimeOffset.UtcNow.Add(lease) });
                return true;
            }
            catch (IOException) when (File.Exists(path))
            {
                var claim = ReadClaim(path);
                if (claim is null || claim.Value.Expires > DateTimeOffset.UtcNow) return false;
                File.Delete(path); // expired: try once more
            }
        }
        return false;
    }

    static (string Owner, DateTimeOffset Expires)? ReadClaim(string path)
    {
        try
        {
            var json = JsonNode.Parse(File.ReadAllText(path))!;
            return (json["owner"]!.GetValue<string>(), json["expires"]!.GetValue<DateTimeOffset>());
        }
        catch (Exception e) when (e is IOException or JsonException or InvalidOperationException)
        {
            return null; // being written by its owner, or already gone
        }
    }

    public void ReleaseClaim(string jobId, string owner)
    {
        string path = Path.Combine(JobDir(jobId), "claim");
        if (ReadClaim(path) is { } claim && claim.Owner == owner) File.Delete(path);
    }

    /// <summary>Writes an output file under a temporary name first: a crash never leaves a half file that looks finished.</summary>
    public async Task<StoredFile> SaveFileAsync(string jobId, string node, string name, byte[] bytes, CancellationToken cancel)
    {
        string dir = Path.Combine(JobDir(jobId), "outputs");
        Directory.CreateDirectory(dir);
        string final = Path.Combine(dir, $"{node}-{Path.GetFileName(name)}");
        string temp = final + ".partial";
        await File.WriteAllBytesAsync(temp, bytes, cancel);
        File.Move(temp, final, overwrite: true);
        return new StoredFile(node, Path.GetFileName(final), bytes.Length, Convert.ToHexStringLower(SHA256.HashData(bytes))[..16]);
    }

    public void Complete(string jobId, string owner, string instance, int attempt, IReadOnlyList<StoredFile> files)
    {
        var done = new JsonObject
        {
            ["job"] = jobId,
            ["worker"] = owner,
            ["gpu"] = instance,
            ["attempts"] = attempt,
            ["files"] = new JsonArray(files.Select(f => (JsonNode)new JsonObject { ["node"] = f.Node, ["name"] = f.Name, ["bytes"] = f.Bytes, ["sha256"] = f.Sha256 }).ToArray()),
        };
        string path = Path.Combine(JobDir(jobId), "done.json");
        File.WriteAllText(path + ".partial", done.ToJsonString(new JsonSerializerOptions { WriteIndented = true }));
        File.Move(path + ".partial", path, overwrite: true);
        ReleaseClaim(jobId, owner);
    }

    public void Fail(string jobId, string owner, string reason)
    {
        File.WriteAllText(Path.Combine(JobDir(jobId), "failed.txt"), reason + Environment.NewLine);
        ReleaseClaim(jobId, owner);
    }
}
